pennylane/optimize/gradient_descent.py

```diff
@@ -47,86 +48,119 @@ def update_stepsize(self, stepsize):
                 objective function with respect to the variables ``x``.
                 If ``None``, the gradient function is computed automatically.
+                Must return the same shape of tuple [array] as the autograd derivative.
+            **kwargs : variable length of keyword arguments for the objective function
 
         Returns:
-            array: the new variable values :math:`x^{(t+1)}`
+            list [array]: the new variable values :math:`x^{(t+1)}`.
+                If single arg is provided, list [array] is replaced by array.
         """
-        g, _ = self.compute_grad(objective_fn, x, grad_fn=grad_fn)
-        x_out = self.apply_grad(g, x)
 
-        return x_out
+        g, _ = self.compute_grad(objective_fn, args, kwargs, grad_fn=grad_fn)
+        new_args = self.apply_grad(g, args)
+
+        # unwrap from list if one argument, cleaner return
+        if len(new_args) == 1:
+            return new_args[0]
+
+        return new_args
 
     @staticmethod
-    def compute_grad(objective_fn, x, grad_fn=None):
+    def compute_grad(objective_fn, args, kwargs, grad_fn=None):
         r"""Compute gradient of the objective_fn at the point x and return it along with the
-            objective function forward pass (if available).
+        objective function forward pass (if available).
 
         Args:
             objective_fn (function): the objective function for optimization
-            x (array): NumPy array containing the current values of the variables to be updated
-            grad_fn (function): Optional gradient function of the objective function with respect to
-                the variables ``x``. If ``None``, the gradient function is computed automatically.
+            args (tuple): tuple of NumPy arrays containing the current parameters for the
+                objection function
+            kwargs (dict): keyword arguments for the objective function
+            grad_fn (function): optional gradient function of the objective function with respect to
+                the variables ``args``. If ``None``, the gradient function is computed automatically.
+                Must return the same shape of tuple [array] as the autograd derivative.
 
         Returns:
-            tuple: The NumPy array containing the gradient :math:`\nabla f(x^{(t)})` and the
+            tuple (array): NumPy array containing the gradient :math:`\nabla f(x^{(t)})` and the
                 objective function output. If ``grad_fn`` is provided, the objective function
                 will not be evaluted and instead ``None`` will be returned.
         """
         g = get_gradient(objective_fn) if grad_fn is None else grad_fn
-        grad = g(x)
+        grad = g(*args, **kwargs)
         forward = getattr(g, "forward", None)
 
         return grad, forward
 
-    def apply_grad(self, grad, x):
-        r"""Update the variables x to take a single optimization step. Flattens and unflattens
+    def apply_grad(self, grad, args):
```

Comment:
Just some [feedback](https://challenge.qhack.ai/jury/clarifications/102) from the hackathon based upon a [submission](https://challenge.qhack.ai/jury/submissions/2246/source) that uses `compute_grad` and `apply_grad`: they had some confusion because they needed to switch from passing `x=params` to `args=[params]` and also (for `compute_grad`) they needed to explicitly pass `kwargs={}`.

Should we consider making these methods be similar to `step(*args, **kwargs)`? Or, alternatively just make these hidden methods.

Reply:
@trbromley I tend to agree, especially if this would be more of a user-facing method (not sure how much of an issue it would be otherwise). There was [a brief discussion re. this](https://github.com/PennyLaneAI/pennylane/pull/959/files/f7e9d67ce2ff6b4a27bace3554bac12b7440753f#r545453403) when this was under review. Perhaps just making them hidden by preceding with an underscore would be good here. 🤔 @albi3ro @josh146

Reply:
I'm somewhat ambivalent here. We definitely considered `compute_grad` and `apply_grad` 'private' methods when first designing the optimizers, as they were not intended to be used as the public optimizer API.

But it probably makes more sense to keep them public, and standardize the signatures, since the fact that users were finding and using them suggests they add value.

Reply:
Unless there's a larger demand from users for this function, it'd be inefficient to unpack and repack the values.   It would add operations for no real purpose.

Reply:
I'm in favour of 'doing nothing' for now unless there is more feedback and demand.